### PR TITLE
🐛 Fix vSphere platform vulnerabilities and eol

### PR DESCRIPTION
**Pull request by czunker**

### CODE REVIEW

**Comment by czunker** [comment]
This sometime works:
```
Asset: 192.168.5.24
-------------------

Data queries:
asset.title: "VMware vSphere 7.0.3"
vsphere.about[fullName]: "VMware vCenter Server 7.0.3 build-18778458"
mondoo.version: "unstable"
platform.vulnerabilityReport: {
  platform: {
    build: "18778458"
    name: "vmware-vsphere"
    release: "7.0.3"
    title: "VMware vSphere 7.0.3"
  }
  published: "2023-10-13T14:19:52Z"
  stats: {
    advisories: {}
... 5 more lines ...
asset.version: "7.0.3"
platform.eol.date: 2025-04-02 02:00:00 +0200 CEST
vsphere.about[osType]: "linux-x64"

Checks:
✓ Pass:         Ensure no known platform advisories exist
✓ Pass:         Ensure no known platform CVEs exist
✓ Pass:         Platform is not end-of-life
✓ Pass:         Ensure the platform is not End-of-Life
```

Sometimes it doesn't:
```
Asset: 192.168.5.21
-------------------

Data queries:
esxi.host.properties[summary][config][product][fullName]: "VMware ESXi 7.0.3 build-19035710"
mondoo.version: "unstable"
error: incorrect provider for asset, not adding
platform.eol.date: null
asset.version: "7.0.3"
esxi.host.properties[hardware][systemInfo][uuid]: "fe664d56-f513-9098-0e12-4e5ffce68410"
esxi.host.properties[hardware][biosInfo][biosVersion]: "VMW71.00V.18452719.B64.2108091906"
error: incorrect provider for asset, not adding
platform.vulnerabilityReport: null
esxi.host.name: "192.168.5.21"
asset.title: "VMware ESXi"
esxi.host.properties[summary][config][product][osType]: "vmnix-x86"

Checks:
! Error:        Ensure no known platform CVEs exist
! Error:        Ensure no known platform advisories exist
! Error:        Platform is not end-of-life
! Error:        Ensure the platform is not End-of-Life
```
This could also differ between assets during the same scan.
When it errors, this is the complete message:
```
! resolver.db> failed to store data, types don't match asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WhkZ26B09ykfl92LPZTjFyQQsR checksum=oeXGLeSaNfw+tjn/M8T10H7AXoZ/tDFdaLtCiTtKFhq9I2QI23dQAa+6acZN3LFRlbX4Ysaf5xTpL/Hk02jkNQ== data={"type":"\u001bplatform.eol"} expected=score received=platform.eol
x failed to send datapoints error="1 error occurred:\n\t* failed to store data for \"oeXGLeSaNfw+tjn/M8T10H7AXoZ/tDFdaLtCiTtKFhq9I2QI23dQAa+6acZN3LFRlbX4Ysaf5xTpL/Hk02jkNQ==\", types don't match: expected score, got platform.eol\n\n"
```

And sometimes it even panics:
```
cnspec scan vsphere christian@vsphere.local@192.168.5.24 --ask-pass --config ~/workspace/mondoo/sa-edge-default.json                                                                                                                                                                                            ✔ │ 21s │ 16:20:53 
Enter password:            
! using builtin provider for vsphere
! using builtin provider for vsphere
→ loaded configuration from /home/christian/workspace/mondoo/sa-edge-default.json using source --config
→ using service account credentials
→ discover related assets for 1 asset(s)
! using builtin provider for vsphere
! using builtin provider for vsphere
! using builtin provider for vsphere
! using builtin provider for vsphere
! using builtin provider for vsphere
→ synchronize assets

 192.168.5.24                    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  42%
... 3 more assets ...

 0/4 scanned                     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  10%

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1df7d6d]

goroutine 361 [running]:
go.mondoo.com/cnquery/v9/providers.(*Runtime).watchAndUpdate(0xc000d64b60, {0xc001857591, 0xc}, {0x0, 0x0}, {0xc001e6ec7c, 0x4}, {0x0?, 0xc0031f1820?})
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:327 +0xcd
go.mondoo.com/cnquery/v9/providers.(*Runtime).WatchAndUpdate(0xc003695940?, {0x2eb3320, 0xc003695de0}, {0xc001e6ec7c, 0x4}, {0xc0051eda80, 0x31}, 0xc0031fe0c0)
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:307 +0xb2
go.mondoo.com/cnquery/v9/llx.runResourceFunction(0xc0031dd0a0, 0xc0031fe030, 0xc0029b0730, 0x100000002)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/builtin.go:744 +0x20a
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).runBoundFunction(0xc0031dd0a0, 0xc0031fe030, 0xc0029b0730, 0x10?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/builtin.go:818 +0xe5
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).runFunction(0xc0031dd0a0, 0xc0029b0730, 0xc005717e60?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:796 +0xd3
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).runChunk(0xc0031dd0a0, 0xc0031dad80?, 0x100000002?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:813 +0x219
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).runRef(0xc001cbddf0?, 0xc0031dad50?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:838 +0xd9
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).runChain(0xc0031dd0a0, 0x2a83850?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:870 +0x9c
go.mondoo.com/cnquery/v9/llx.(*blockExecutor).run(0xc0031dd0a0)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:354 +0x2df
go.mondoo.com/cnquery/v9/llx.(*MQLExecutorV2).Run(0xc001cc0c00?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:286 +0x53
go.mondoo.com/cnspec/v9/policy/executor/internal.(*executionManager).executeCodeBundle(0xc002bb8cd0, 0xc002398460, 0x10?, {0x0, 0x0})
	/home/christian/workspace/mondoo/github.com/cnspec/policy/executor/internal/execution_manager.go:164 +0x4e8
go.mondoo.com/cnspec/v9/policy/executor/internal.(*executionManager).Start.func1()
	/home/christian/workspace/mondoo/github.com/cnspec/policy/executor/internal/execution_manager.go:85 +0x1cb
created by go.mondoo.com/cnspec/v9/policy/executor/internal.(*executionManager).Start in goroutine 161
	/home/christian/workspace/mondoo/github.com/cnspec/policy/executor/internal/execution_manager.go:54 +0x65
```

**Comment by czunker** [comment]
I also tried adding the `extend` keyword to the resources, but this makes it worse:
```
Asset: 192.168.5.24
-------------------

Data queries:
asset.version: "7.0.3"
vsphere.about[fullName]: "VMware vCenter Server 7.0.3 build-18778458"
error: incorrect provider for asset, not adding
platform.vulnerabilityReport: null
error: incorrect provider for asset, not adding
platform.eol.date: null
mondoo.version: "unstable"
vsphere.about[osType]: "linux-x64"
asset.title: "VMware vSphere 7.0.3"

Checks:
! Error:        Ensure the platform is not End-of-Life
! Error:        Ensure no known platform CVEs exist
! Error:        Ensure no known platform advisories exist
! Error:        Platform is not end-of-life
```

I allow to add the provider, I see this error instead:
```
error: no connection to provider
```
But still, all errors.

And looking closer at the code, extended resources are still handled by the root provider. The schema can hold a resource only once. That might also explain the nondeterministic behavior in the previous iteration. Depending on the order the providers and their resources got added to the schema, it works or not.

**Comment by czunker** [comment]
I have one more rough idea in my head, which I need to sketch out. Perhaps this can work with imports for lr and adjusted os/vsphere providers.

**Comment by czunker** [comment]
With the latest changes, I get consistent results for the scans:
```
cnspec scan vsphere christian@vsphere.local@192.168.5.24 -p XXX --config ~/workspace/mondoo/sa-edge-default.json                                                        
→ loaded configuration from /home/christian/workspace/mondoo/sa-edge-default.json using source --config
→ using service account credentials
→ discover related assets for 1 asset(s)
→ synchronize assets

 192.168.5.24                    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% score
 192.168.5.21                    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% score
 192.168.5.22                    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% score
 192.168.5.23                    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% score

 4/4 scanned                     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%

Asset: 192.168.5.21
-------------------

Data queries:
platform.vulnerabilityReport: {
  advisories: [
    0: {
      ID: "VMSA-2022-0025"
      Mrn: "//vadvisor.api.mondoo.app/advisories/VMSA-2022-0025"
      cves: [
        0: {
          ID: "CVE-2022-31680"
          Mrn: "//vadvisor.api.mondoo.app/cves/CVE-2022-31680"
          cvss: [
... 391 more lines ...
platform.eol.date: 2025-04-02 02:00:00 +0200 CEST

Checks:
✕ Fail:         Ensure no known platform CVEs exist
✕ Fail:         Ensure no known platform advisories exist
✓ Pass:         Platform is not end-of-life
✓ Pass:         Ensure the platform is not End-of-Life

...

Asset: 192.168.5.24
-------------------

Data queries:
platform.vulnerabilityReport: {
  platform: {
    build: "18778458"
    name: "vmware-vsphere"
    release: "7.0.3"
    title: "VMware vSphere 7.0.3"
  }
  published: "2023-10-16T10:56:47Z"
  stats: {
    advisories: {}
... 5 more lines ...
platform.eol.date: 2025-04-02 02:00:00 +0200 CEST

Checks:
✓ Pass:         Platform is not end-of-life
✓ Pass:         Ensure no known platform advisories exist
✓ Pass:         Ensure the platform is not End-of-Life
✓ Pass:         Ensure no known platform CVEs exist


Scanned 4 assets

VMware ESXi
    C 192.168.5.21
    A 192.168.5.22
    C 192.168.5.23

VMware vSphere 7.0.3
    A 192.168.5.24

Summary
=======

Score Distribution		Asset Distribution
------------------		------------------
A   2 assets      		VMware ESXi            3
B   0 assets      		VMware vSphere 7.0.3   1
C   2 assets
D   0 assets
F   0 assets

See more scan results and asset relationships on the Mondoo Console: 
```

File: providers/github/go.mod [hunk 1]
```diff
@@ -33,7 +33,7 @@ require (
 	github.com/OpenPeeDeeP/depguard/v2 v2.1.0 // indirect
 	github.com/ProtonMail/go-crypto v0.0.0-20230923063757-afb1ddc0824c // indirect
 	github.com/StackExchange/wmi v1.2.1 // indirect
-	github.com/alecthomas/participle v0.7.1 // indirect
+	github.com/alecthomas/participle v0.3.0 // indirect
```
Comment (czunker) [inline providers/github/go.mod:36]: I needed this downgrade to fix https://github.com/mondoohq/cnquery/issues/2203

**Comment by chris-rock** [comment]
superseeded by #2301

**Pull request closed without merging**